### ISSUE TITLE
Dynamic import fails in polyfill mode for import-maps based project

### ISSUE THREAD

**Issue by tmsns** [guybedford/es-module-shims#333]
Hi, we are trying to switch our import-maps based project from SystemJS to es-module-shims. Unfortunately, we are encountering an import failure in non-Chrome browsers.

These are the conditions of the failure:
- es-module-shims in polyfill mode (we want to benefit from the performance benefits of the polyfill mode)
- only happens on browsers that have no built-in import-maps support (eg. Firefox)
- import failure occurs on a dynamic import coming from the entry module 
(the dynamic import is implemented using `importShim` as advised [here](https://github.com/guybedford/es-module-shims#polyfill-edge-case-dynamic-import))
- the dynamic import has a static import on the entry module  
(Rollup has bundled the common functionality in the entry module)

### Current behavior: 
The entry module is reexecuted. Because of a side effect in the entry module that can only run once (`customElements.define`) the import throws and the dynamic import fails.

### Expected behavior: 
The entry module is executed only once (as it happens in Chrome) and the dynamic import succeeds.

### Minimal repro
https://github.com/tmsns/esm-shims-polyfill-issue

### Some thoughts
- It seems to be a combination of the two edge cases of the polyfill mode: "dynamic import" and "instance sharing"
- I've been looking into the code to see what it would take to make the dynamic import work. I was thinking of making the dynamic import depend on the actual module instead of converting every of its dependencies to a blob. When doing some tests, it *seemed* to be working, but I'm not really sure whether this solution  could have any other downsides.

**Comment by guybedford** [comment]
The standard technique here is to always use an import map to load the entry point of the application itself:

```html
<script type="module">import 'app'</script>
```

Or at least to ensure that there are import maps used in the top-level static graph.

Double execution is unavoidable for the polyfill mode, so you want to ensure the static graph will error when import maps are not supported by using them in the static graph.

**Comment by tmsns** [comment]
Thanks for the feedback!

Unfortunately, using the standard technique on the entry module will delay the execution of the module considerably, as it will make the module go through the blob transformation.

For the second suggestion, the default behavior of Rollup is to bundle the top-level static graph in one chunk. It should be possible to split that up. But that new chunk with the common code (that is needed when rendering the initial parts of the page) will on its turn also go through the blob transformation, because of the import map.

We are trying to be on par with the native module performance or performance similar to SystemJS. With that in mind, I went a bit deeper in the idea of reusing modules that already ran, as the initial poc seem to work in Firefox. I was able to make the solution more generic, and this is what I came up with:
https://github.com/tmsns/esm-shims-polyfill-issue/compare/main...fix/use-native-module

Let me know what you think of this approach. In the mean time, I am gonna do some more in-depth tests with other browsers/setups.

**Comment by thepassle** [comment]
To clarify, here is an example of the issue we're seeing:

In our (@tmsns and me) usecase, we have a multi page app, that loads esm-shims, and an app shell custom element and some static HTML, the code looks as follows:
![Screenshot 2022-09-01 at 14 32 29](https://user-images.githubusercontent.com/17054057/189677479-9b684494-32b0-438f-b4d8-83aa27f00ce4.png)

Given this example, we see the following behavior in the browser:

https://user-images.githubusercontent.com/17054057/189677605-567b2aab-2958-400b-8762-bbe661bd4a60.mov

It's very faint, but there's a slight 'flicker' in the loading of the app-shell module, on subsequent refreshes this flicker is still visible (presumably because the app-shell module has been turned into a blob, and so instead of loading it from the browser cache, it's being loaded by esm-shims)

You can see the difference with the current situation using systemjs:
![Screenshot 2022-09-01 at 14 32 02](https://user-images.githubusercontent.com/17054057/189677948-742a3b16-bbe1-4769-b064-58e4f7f2ab60.png)

https://user-images.githubusercontent.com/17054057/189678003-04f71c49-1cff-4304-82a8-403ff18dd5fb.mov

In a way, the flicker makes sense, and it sort of seems like esm-shims is doing exactly what it's intended to do, but what we're trying to figure out is if there is any way to avoid the flicker. It may seem very minimal in the reproduction video I posted above, but in the real application the effect is definitely very noticeable and disruptive:

This is using esm-shims:

https://user-images.githubusercontent.com/17054057/189679872-8c67d544-2897-41be-adad-b51f2ed751d5.mov

This is using systemjs:

https://user-images.githubusercontent.com/17054057/189679921-6f713b29-be3e-48ab-9d3f-4acac8426999.mov


Do you by any chance have any further insights on if there's any kind of way to avoid the flicker effect, @guybedford ?

You can find the reproduction code here if needed 🙂: https://github.com/thepassle/system-esm-shims-repro

**Comment by guybedford** [comment]
It might help to check some timing logs around why this is such a big difference between SystemJS and ES Module Shims, because it really shouldn't be.

The blob technique is very fast, so I'm quite surprised that is enough to make a difference here.

Some other things to check:
* DOMContentLoaded timing
* ready state timing
* ES Module shims init timing (the initPromise in https://github.com/guybedford/es-module-shims/blob/main/src/es-module-shims.js#L122)
* Sub-phases of ES module shims init can even be investigated, such as the feature detection phase which can cause a delay in browsers other than Chrome (https://github.com/guybedford/es-module-shims/blob/main/src/features.js#L15)
* Setting `noLoadEventRetriggers` - https://github.com/guybedford/es-module-shims#no-load-event-retriggers, this could sometimes cause a flash if there are DOMContentLoaded or readystate attachments which are being re-run
* Another thing might be to try a top-level `<script>importShim('entry')</script>` directly after ES Module Shims to see if that affects timing

As I say of all the above, the use of blobs sounds like the least possible cause here, but if it really is that that is definitely worth investigating too.

Would be great to dive into this further definitely, just let me know if I can assist some more.

**Comment by thepassle** [comment]
> As I say of all the above, the use of blobs sounds like the least possible cause here, but if it really is that that is definitely worth investigating too.

My reasoning for this behavior is that it's just the difference between how the scripts/module scripts are loaded. In the systemjs example, the `app-shell.js` file is loaded as a regular script tag which blocks rendering: 

![image](https://user-images.githubusercontent.com/17054057/189701708-f9084b42-b7ff-4a38-a320-bacf87a5e233.png)

When using native ESM, the `app-shell.js` module script is `defer`ed ([mdn](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-defer)). With native ESM, on subsequent refreshes `app-shell.js` can be loaded from the cache. When using es-module-shims, the `app-shell.js` module script is deferred, _then_ loaded via a blob. On subsequent requests, those blobs won't be loaded from the cache.

This is consistent with the behavior I see in the reproduction repo:

## Using native esm:

- initial load has the flicker
- on refresh there is no flicker and `app-shell.js` loads from the cache 

> (note: caching in devtools is enabled)

https://user-images.githubusercontent.com/17054057/189700963-258b4905-1745-4408-b5d4-160b29644c03.mov

## Using es-module-shims:

- initial load has flicker
- subsequent reloads also have a flicker 

> (note: caching in devtools is enabled)

https://user-images.githubusercontent.com/17054057/189701282-498e762a-5919-41ce-ae1f-7981477423ab.mov

**Comment by guybedford** [comment]
@thepassle cache behaviour can't be relied upon though - the standard render designed for should be a cold load. If it really is just down to the async / blocking question, then yes, native ESM has defer semantics and all applications using native ESM have to be designed to deal with that!

**Comment by guybedford** [comment]
Another thing worth noting here - did you try adding the `async` modifier to the module script? That should disable the default defer semantics allowing the module to execute before DOMContentLoaded. ES Module Shims will respect these semantics as well, where a `<script type="module" async>` will actually execute earlier than a `<script type="module">` in polyfill mode.

**Comment by tmsns** [comment]
First off, thanks for your feedback, @guybedford. 

We went a bit deeper into the different proposals and created a page for each case:
https://github.com/tmsns/esm-shims-polyfill-issue/tree/cases

Here are the results:
![image](https://user-images.githubusercontent.com/17364880/191025871-a9db8ad7-cb11-4ef4-ae19-5b5d1be5b466.png)
- We tested on Firefox with GPRS throttling on and compared cold loads (cache disabled) with warm loads (cache enabled).
  In our use case we have a MPA that triggers full page reloads when the user navigates from one screen to another. This means the first hit is cold, but every subsequent one is a hot one. Being able to improve the perceived performance during the user's journey on these different pages greatly improves their experience.
- **load entry point using an import**
  This solves the initial issue, but introduces more heavy flickering in the cached version.
- `noLoadEventRetriggers`
  This technique does not seem to stop the flickering, even with caching.
- **async module script**
  This technique does not seem to change the experience as opposed to without the attribute: both have almost no flickering on hot loads.  But as this enables the polyfill mode, we have the same problem as the initial issue: the dynamic import fails because of the double execution.

The only case that seems to give the best result is the one based on the **polyfill mode where we changed how dependencies get loaded**:
- As we use the polyfill approach, when caching is active, there is rarely flickering.
- The change in dependency loading allows the reuse of the originally loaded module instead of reloading it as a blob. Next to fixing the original issue, **it also improves performance as the entry module is only loaded once and is reused within the blob modules**.

This is what the change looks like. Until now, we haven't seen any side effect:
![image](https://user-images.githubusercontent.com/17364880/191033264-f8f19e51-089c-432d-8ed8-a0caca4d8c22.png)

**Comment by guybedford** [comment]
@tmsns thanks for taking the time to clearly explain the cases here, that helps a lot to understand what you're dealing with.

Firstly, in the polyfill case, app-shell.js is being executed twice. The reason for this is that the polyfill does not reuse modules from the native loader. And the reason for this is that doing that results in double fetches since Firefox and Safari don't share their internal fetch caches with the module loader even for the same exact fetch options.

For this reason, in the polyfill case I would recommend using a direct script tag rather than a module, since you are always executing that code natively in all environments. When using polyfill mode you want all modules to use import maps to go through polyfill mode for the above reasons. This is explained in the readme, perhaps it could be worded better.

It seems to make sense that the async case likely only applies to very slow DOM rendering cases where the scripts are at the top of the HTML, so that effect does seem negligible here.

For the fix, basically what you are doing with that approach is telling es-module-shims not to process the URL at all if it corresponds to a script tag, which effectively would disable polyfilling so that's not something that can be upstreamed.

But the approach you're using is already provided as a feature of the project, via the `skip` (regex) option:

```html
<script type="esms-options">
{
  "skip": "app-shell\.js$"
}
</script>
```

The above will do exactly what you've written in the patch while also stopping the double execution from happening in the polyfill case.

The regex is a bit difficult to use though, I've posted a PR at https://github.com/guybedford/es-module-shims/pull/338 with an update to support a direct array of modules to skip as unnormalized URLs, so you could just write `"skip": ["./app-shell.js"]`. I tested that build on the polyfill version and it seems to work well to me.

Just let me know if I've missed anything else on this topic.

**Comment by tmsns** [comment]
For you as well, @guybedford, your investigations and proposals help a lot! Thanks! :-)

> For this reason, in the polyfill case I would recommend using a direct script tag rather than a module, since you are always executing that code natively in all environments. When using polyfill mode you want all modules to use import maps to go through polyfill mode for the above reasons. This is explained in the readme, perhaps it could be worded better.

That was one of our first setups. Unfortunately, it broke/brakes our use-case.
We need app-shell as a module, as it exports some of the (bundled) common functionality. In the repro, the dynamic import (`foo.js`) has a dependency on it. So, after the file has run as a direct script, `foo.js` will try to import it, but this time as a module. The file will go through the blob transformation, but on execution it will break as the side-effect of defining the custom element (again) will throw.


> But the approach you're using is already provided as a feature of the project, via the `skip` (regex) option:
> 
> ```
> <script type="esms-options">
> {
>   "skip": "app-shell\.js$"
> }
> </script>
> ```
> 
> The above will do exactly what you've written in the patch while also stopping the double execution from happening in the polyfill case.
> 
> The regex is a bit difficult to use though, I've posted a PR at #338 with an update to support a direct array of modules to skip as unnormalized URLs, so you could just write `"skip": ["./app-shell.js"]`. I tested that build on the polyfill version and it seems to work well to me.

This is exactly what we were trying to achieve: having the app-shell/entry module go through the native passthrough. I had not noticed the `skip` option before. But, first tests show that it is indeed working the way you explained. The dynamic import is working in Firefox and perceived performance seem ok. :-)
The fix I had written automatically made it use the native passthrough, but passing in options (probably I'll use the JS method `window.esmsInitOptions`) should still be doable.
My next step now is to implement this in the actual project.
If I notice anything else, I'll let you know!

PS: just in case, I've updated the repro to showcase your two new proposals

**Comment by guybedford** [comment]
@tmsns 1.6.0 has been released with support for an array of paths or relative paths in `skip`. Do report back if anything else comes up, otherwise I'm closing this issue for now.
